src/machine.rs

```diff
@@ -308,6 +313,65 @@ impl<'mir, 'tcx: 'mir> PrimitiveLayouts<'tcx> {
     }
 }
 
+#[derive(Clone, Debug)]
+pub struct Mapping {
+    pub range: Range<Size>,
+    pub alloc_id: AllocId,
+    pub can_read: bool,
+    pub can_write: bool,
+}
+
+impl Mapping {
+    pub fn contains(&self, addr: Size) -> bool {
+        self.range.contains(&addr)
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct Mappings {
```

Comment:
This datastructure is scary. It seems to have a lot of invariants that I can't quite make heads or tail of. I'd hope we could reuse an existing datastructure instead of rolling our own thing.

Reply:
I agree. I'm trying to rewrite this based on `RangeMap`, and now I'm tied up in the strange semantics that the documentation for `mremap` prescribes. It refers to "a mapping", where `mprotect` simply refers to pages. And it's hard to see how to do things with `RangeMap` on a _mapping_ as opposed to just... memory.

Reply:
Okay well example programs in combination with documentation help a lot.

Consider the following example. This maps 3 pages, then does a remap which includes the first 2 pages, but remaps down to the size of one page. So of the 3 pages we initially mapped, only the first and last are mapped, and the middle page is not mapped. 
```rust
fn main() {
    unsafe {
        let ptr = libc::mmap(
            std::ptr::null_mut(),
            4096 * 3,
            libc::PROT_READ | libc::PROT_WRITE,
            libc::MAP_PRIVATE | libc::MAP_ANONYMOUS,
            -1,
            0,
        );
        println!("{:p}", ptr);

        let res = libc::mremap(ptr, 4096 * 2, 4096, 0);
        assert_eq!(res, ptr);
        println!("{:?}", std::io::Error::last_os_error());

        println!("{}", *(ptr.add(4096 * 0) as *const u8));
        //println!("{}", *(ptr.add(4096 * 1) as *const u8)); // SIGSEGV
        println!("{}", *(ptr.add(4096 * 2) as *const u8));
    }
}```
This is... spooky, to say the least. If we are modeling the `mmap` as an allocation, we now have a discontinuous region of address space with the same AllocId. That doesn't seem right.